c10/core/Scalar.h

```diff
@@ -24,6 +24,11 @@ namespace c10 {
  */
 class C10_API Scalar {
  public:
+  // constructor tag used by empty scalar constructor
+  struct MakeEmptyScalar {};
+
+  explicit Scalar(MakeEmptyScalar) : tag(Tag::HAS_n) {}
```

Comment:
I borrowed the tagged constructor strategy from `intrusive_ptr` 😋 

On second thought, these really should be private, and `OptionalScalarRef` should be a `friend` of `Scalar`.